### PR TITLE
Date: Add timezone hint

### LINKED ISSUE #15673: Add timezone to Publish date/time for scheduled posts
**Describe the bug**

For multi users blog (in multiple time zones) it is not clear what time zone a scheduled post is being set to. It appears the time shown for the Publish date is the time zone for the blog and not the user.

Showing the time zone used in the displayed date will help clarify for all users of the site.

image

### PR DESCRIPTION
**Pull request by obenland**
## Description
This will add timezone information to the schedule post UI for users whose local timezone differs from the timezone set for the site. It shows the timezone abbreviation and a tooltip on hover with the full timezone name and UTC offset.

Fixes #15673.

## How has this been tested?
Tested locally with various timezones set in wp-admin.

To test:
* Got to Settings > General and select a timezone for your site.
* In the "Document" sidebar panel, click on the publish date to bring up the publish calender.
* Make sure the timezone information at the end of time component only gets displayed when your local timezone is different from the site's timezone.

## Screenshots 
82586619-b386d800-9b4c-11ea-8770-55bec0448124

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by retrofox** [comment]
Getting some issues when testing with some timezones:

```
react_devtools_backend.js:6 Moment Timezone has no data for America/Argentina/Buenos_Aires. See http://momentjs.com/timezone/docs/#/data-loading/. 
```

image

I think it produces the `UTC` that doesn't render properly?

Screen Shot 2020-06-23 at 4 08 06 PM

I'd expect to see `UTC-3` below, in the red square.

**Comment by obenland** [comment]
Yes, this PR relies pretty heavily on the data that is bundled with Moment. For your timezone [it doesn't seem to contain proper timezone abbreviations](https://github.com/moment/moment-timezone/blob/c3e6e4f550638f5c8fcbc7321b4ab4cedac855b0/data/packed/latest.json#L28). 

I've not come across the "Moment Timezone has no data" error before. Does that happen for all timezones you tested with?

**Comment by retrofox** [comment]
> I've not come across the "Moment Timezone has no data" error before. Does that happen for all timezones you tested with?

It happens for all timezones. Maybe because I'm testing with WP 5.3?

image

**Comment by jasmussen** [comment]
Thanks for the PR! 

Before:

Screenshot 2020-06-24 at 08 15 12

After:

Screenshot 2020-06-24 at 08 15 51

Don't mind the visual changes, a good rebase should take care of that. 

Nice, this does seem to show the correct UTC for me.

I noticed that the entire window closes as soon as you click AM/PM or a date. That's not a regression of this PR and exists in the master branch, but seems like a regression nonetheless, right?

close

Functionality wise, and UI wise, this looks fine enough. Though we probably need a way for a keyboard user to learn this information. It looks like it's enough to simply add `tabindex="0"` to the div holding the UTC information, but we might need a role too. CC: @enriquesanchez in case you have anything to add here.

File: packages/components/src/date-time/time.js [hunk 1]
```diff
@@ -277,6 +279,35 @@ class TimePicker extends Component {
...
+					: 'UTC' + offset;
+
+			return (
+				<Tooltip position="top center" text={ timezoneDetail }>
+					<div className="components-datetime__timezone">
+						{ timezoneAbbr }
+					</div>
+				</Tooltip>
+			);
+		};
```
Comment (marekhrabe) [inline packages/components/src/date-time/time.js:309]: Unless I've missed something, this looks pretty standalone. Let's pull the declaration out of the `render` function

**Comment by enriquesanchez** [comment]
I'm not sure what role would be appropriate for this piece of information. I'm also not sure if tabbing to it is the right approach, given that the user can't really interact with this content, it's purely informational.

Let me do some research and ask around and I will get back to you ASAP.

**Comment by enriquesanchez** [comment]
>I'm also not sure if tabbing to it is the right approach, given that the user can't really interact with this content, it's purely informational.

Looks like I was wrong here. [I asked for some feedback from the Accessibility team](https://wordpress.slack.com/archives/C02RP4X03/p1593031760489200), and looks like this element can be focusable, as Joen suggested, because the tooltip _is_ the interaction.

>Note that when I hear "tooltip on hover" I immediately red flag that. That would actually mean that it should be focusable, and the focus should give access to the tooltip. In that case, it should not be tied via aria-describedby, it should just be a tabstop where the timezone is disclosed.

>It is interactive if there's a keyboard-accessible tooltip associated with it. The interaction is that on focus, further information is disclosed.

Thanks to @joedolson for the feedback and help.

**Comment by afercia** [comment]
After some conversation on Slack and after reading the related issue, I'm not fully sure this feature should be in core.

Quoting form the issue:
> For multi users blog (in multiple time zones) it is not clear what time zone a scheduled post is being set to.

This doesn't appear to be a need for the vast majority of wp.org users. It is a need for a very specific group of users. This fact alone sounds like it should be _plugin territory_.

One of the points on the related issue #15673 mentions this as well:
> (5) Don't change in core or Editor and leave for a plugin to change ...

Given the above, and given the accessibility concerns discussed on Slack, I'd highly recommend to ship this feature in a plugin so that it can be used where it's really relevant: multi users blogs.

**Comment by obenland** [comment]
Thank you for that feedback—I agree, it's not super useful for users that are in the same timezone as the site they're working on.

I updated the PR to only display timezone information when the user timezone differs from the site timezone. That way it only gets displayed for users where that information is relevant.

**Comment by jeffpaul** [comment]
As someone who regularly schedules posts on both multi-user and single-user sites and who gets frustrated not knowing when the post is really scheduled for, seeing this PR land would make my life much easier when it comes to scheduling content.  I have single-user sites that I've set up over time and I've moved across multiple timezones, so I always forget what timezone I may have set a site up on.  And contrary to the latest commit on this PR, even if my local timezone matches that of the site I'd still want to see what timezone is being used as a comforting reminder that the time I'm scheduling in fact does match my local timezone.

IOU a round of tacos and preferred paired drink for getting this into core.

**Comment by obenland** [comment]
@ajitbohra @chrisvanpatten @youknowriad Do you have any advice on next steps for this PR? is there anyone I should ping for review?

**Comment by obenland** [comment]
Is there anything I can do to encourage a review of this PR?

**Comment by draganescu** [comment]
I rebased this PR and tried to test it locally but for some reason the TZ indicator doesn't show up:

Screenshot 2020-07-07 at 19 29 08

I did set WP timezone to be different from system. What am I doing wrong?

**Comment by obenland** [comment]
@draganescu When you select one of the UTC+/-X options in the timezone dropdown it currently doesn't display the timezone info since there is no additional info to populate the popup.

**Comment by obenland** [comment]
I'm now running into [the same error as @retrofox earlier](https://github.com/WordPress/gutenberg/pull/23400#issuecomment-648362668). Is there anything I can do to make sure the timezone information is bundled and available? @mkaz Do you know by chance?

**Comment by jeffpaul** [comment]
@obenland any chance that's affected by core bumping moment.js?  https://core.trac.wordpress.org/ticket/50408

**Comment by mkaz** [comment]
I'm looking now. There is a second npm package called [moment-timezone](https://www.npmjs.com/package/moment-timezone) that I believe is also required, I'm not sure how it all gets bundles and included.

**Comment by youknowriad** [comment]
Maybe related but for performance reasons, we don't bundle "moment-timezone" entirely in Gutenberg/Core. That package is too big and we only need the timezone info provided by the WP server.

**Comment by mkaz** [comment]
Ok this diff kinda works, but I see a lot of errors and this does include the moment-timezone which looks like a non-starter. 

At least this may help troubleshoot. I remember that moment does some weird stuff with globals, and just include the timezone lib tweaks things and there is conflict with core moment.

```diff
diff --git a/package-lock.json b/package-lock.json
index b0cf3da793..e7e53f1867 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -11099,6 +11099,7 @@
                                "@emotion/styled": "^10.0.23",
                                "@wordpress/a11y": "file:packages/a11y",
                                "@wordpress/compose": "file:packages/compose",
+                               "@wordpress/date": "file:packages/date",
                                "@wordpress/deprecated": "file:packages/deprecated",
                                "@wordpress/dom": "file:packages/dom",
                                "@wordpress/element": "file:packages/element",
diff --git a/packages/components/src/date-time/timezone.js b/packages/components/src/date-time/timezone.js
index 14e36caf95..84097ae2a5 100644
--- a/packages/components/src/date-time/timezone.js
+++ b/packages/components/src/date-time/timezone.js
@@ -1,7 +1,9 @@
 /**
  * External dependencies
  */
-import moment from 'moment';
+import momentLib from 'moment';
+import 'moment-timezone/moment-timezone';
+import 'moment-timezone/moment-timezone-utils';

 /**
  * WordPress dependencies
@@ -19,7 +21,7 @@ import Tooltip from '../tooltip';
  */
 const TimeZone = () => {
        const { timezone } = getDateSettings();
-       const userTimezone = moment.tz.guess( true );
+       const userTimezone = momentLib.tz.guess( true );

        if (
                ! timezone.string ||
@@ -31,7 +33,7 @@ const TimeZone = () => {

        const offset =
                timezone.offset > 0 ? '+' + timezone.offset : timezone.offset;
-       const zoneAbbr = moment.tz( new Date(), timezone.string ).zoneAbbr();
+       const zoneAbbr = momentLib.tz( new Date(), timezone.string ).zoneAbbr();
        const timezoneDetail =
                'UTC' === timezone.string
                        ? __( 'Coordinated Universal Time' )
```

**Comment by mkaz** [comment]
ok, my diff isn't a great one, it works in some regard but not for the right reason.

It seems the result of guessing the user timezone is not accurate: `moment.tz.guess( true );`

The existing function is returning `WP` and thus bypassing the tool tip, since there is a explicit check for that.

The diff I made above overrides the returning of `WP` but it is returning `null` instead of a valid guess for the user's timezone. It looks like this is being interpreted as UTC-0 and thus shows if the WP timezone does not equal UTC-0.

I think what we need is a `moment.tz.guess()` or similar function to get the proper user's timezone and compare that offset to the timezone.offset and not the string compare.

**Comment by mkaz** [comment]
Still researching, looks like @ryelle PR here is related: https://github.com/WordPress/gutenberg/pull/22866

**Comment by obenland** [comment]
[Loading a 10-year version of Moment's timezone data](https://github.com/WordPress/gutenberg/compare/add/timezone-tooltip...try/add-timezone-data) worked well for me. I also brought it up as a possibility in @ryelle's PR.

This PR is definitely depended on the outcome of #22866.

**Comment by mkaz** [comment]
ok, I can commit this to the branch, but don't want to step on any toes, 
I think this patch will do what we want and remove the dependency on Moment.

One thing lost is the zoneAbbr that is nice, mapping UTC-7 to PDT, but I think we can get that a different way without requiring moment. I think we could add that zone abbreviation to `getDateSettings()` in date package, the string and offset are already there.

Take a look and let me know what you think...

```diff
diff --git a/packages/components/src/date-time/timezone.js b/packages/components/src/date-time/timezone.js
index 14e36caf95..96f68eee89 100644
--- a/packages/components/src/date-time/timezone.js
+++ b/packages/components/src/date-time/timezone.js
@@ -1,8 +1,3 @@
-/**
- * External dependencies
- */
-import moment from 'moment';
-
 /**
  * WordPress dependencies
  */
@@ -19,31 +14,27 @@ import Tooltip from '../tooltip';
  */
 const TimeZone = () => {
        const { timezone } = getDateSettings();
-       const userTimezone = moment.tz.guess( true );

-       if (
-               ! timezone.string ||
-               timezone.string === userTimezone ||
-               'WP' === userTimezone
-       ) {
+       // getTimezoneOffset() returns in minutes and opposite how you'd expect
+       // America/Los_Angeles = 420 and not -7
+       const userTimezoneOffset = -1 * ( new Date().getTimezoneOffset() / 60 );
+
+       // System timezone and user timezone match, nothing needed
+       if ( timezone.offset === userTimezoneOffset ) {
                return null;
        }

-       const offset =
-               timezone.offset > 0 ? '+' + timezone.offset : timezone.offset;
-       const zoneAbbr = moment.tz( new Date(), timezone.string ).zoneAbbr();
+       const offsetSymbol = timezone.offset >= 0 ? '+' : '';
+
+       const zoneAbbr = `UTC${ offsetSymbol }${ timezone.offset }`;
        const timezoneDetail =
                'UTC' === timezone.string
                        ? __( 'Coordinated Universal Time' )
-                       : '(UTC' + offset + ') ' + timezone.string.replace( '_', ' ' );
-       const timezoneAbbr =
-               '' !== timezone.string && isNaN( zoneAbbr ) ? zoneAbbr : 'UTC' + offset;
+                       : `(${ zoneAbbr }) ${ timezone.string.replace( '_', ' ' ) }`;

        return (
                <Tooltip position="top center" text={ timezoneDetail }>
-                       <div className="components-datetime__timezone">
-                               { timezoneAbbr }
-                       </div>
+                       <div className="components-datetime__timezone"> { zoneAbbr } </div>
                </Tooltip>
        );
 };
```

**Comment by mkaz** [comment]
Ok, after taking a deeper look and going to the source of the system timezone, it is easier to export the timezone abbreviation from PHP, than trying to recreate using Moment.

I committed this change https://github.com/WordPress/gutenberg/pull/23400/commits/9ed1320a7cdcd1f5fcc790ac3790e73cf2a8afdc that updates the date settings ouput via inline-script to include a new `abbr` property. The abbreviated timezone is calculated using PHP date/time functions.

This removes the need to include moment for this component.

It works if the string is set to a city/locale, for example `America/New_York`, or `America/Los_Angeles` showing the appropriate abbreviated timezone (EDT or PDT). If a UTC time is set, for example UTC-7, it will show that as the abbreviated value, because the timezone set to UTC-7 is not enough info to know the short timezone, for example: is it because it is pacific daylight time, or mountain regular time.

I believe this does what we want. Please test it out by changing your WordPress timezone setting, if it differs from your user timezone (as defined by your browser) than it will show the tool tip.

Thanks @obenland for continuing the effort here and the ping, this has been a long-standing issue I've wanted to resolve. A bummer we missed WP 5.5, I'm sorry I didn't see it earlier, but at least it'll finally get closed.

Let me know what y'all think, it feels like a better approach.

**Comment by obenland** [comment]
Nice work @mkaz! This works nicely for me. Feels much better without Moment.

Do the changes in `gutenberg_add_date_settings_timezone()` need to be ported over into Core in a separate ticket?

**Comment by mkaz** [comment]
> Do the changes in `gutenberg_add_date_settings_timezone()` need to be ported over into Core in a separate ticket?

Yes, but I think that will only done at merge, so wouldn't be until WP 5.6, I can create the ticket for it in Trac. I wanted to wait until we decided this was a good approach or not.

**Review by mkaz — APPROVED** [review]
It feels a little off approving, since I contributed a chunk of the code, but considering @obenland reviewed my piece and I his, it feels like it should be legit.

We can see if anyone else has any thoughts and merge tomorrow if nothing comes up.

**Review by draganescu — APPROVED** [review]
Tested this locally and now it works as described.

**Comment by jasmussen** [comment]
Thank you Andrei 💯

**Comment by mkaz** [comment]
Associated Trac ticket for core: https://core.trac.wordpress.org/ticket/50624

**Comment by blogjunkie** [comment]
Thanks @mkaz for adding this enhancement, but I feel it's still really confusing. Here's a screenshot from a blog that has it's timezone set to PST and the user scheduling the post is in EST. The user is scheduling the post for 5am PST, which is 8am EST in her own timezone. The UI shows 5am and 8am together, which is confusing. 

calendar_time1

I'd like to suggest that we simplify this by not showing the user's local time at all. This will avoid confusion of when the post will be published, and negates the need for adding the timezone hint in the first place. Thank you for considering this feedback and suggestion.

**Comment by mkaz** [comment]
I agree its still confusing. 

This change is not yet available in WordPress core, it is slated for WP 5.6. So, what you are showing does not have what this PR is trying to fix, what we hope will make it less confusing.

The interface should be only showing the server time and a hint at what that time is, I don't see that hint in your screenshot, for you it should be showing PST (though really it would be PDT because its daylight saving time right now) 

However, in testing, it does look like there is a bug and mine is showing my current time but with the label of the server which is incorrect. My local setup is opposite, I'm in PDT and the site is set to EDT.

date-time

I created an issue to address the bug in #25256

**Pull request merged**